src/core/utils/qfieldcloudutils.cpp

```diff
@@ -86,3 +89,70 @@ const QVariant QFieldCloudUtils::projectSetting( const QString &projectId, const
   const QString projectPrefix = QStringLiteral( "QFieldCloud/projects/%1" ).arg( projectId );
   return settings.value( QStringLiteral( "%1/%2" ).arg( projectPrefix, setting ), defaultValue );
 }
+
+const QMultiMap<QString, QString> QFieldCloudUtils::getPendingAttachments()
+{
+  QMultiMap<QString, QString> files;
+
+  QLockFile attachmentsLock( QStringLiteral( "%1/attachments.lock" ).arg( QFieldCloudUtils::localCloudDirectory() ) );
+  if ( attachmentsLock.tryLock( 10000 ) )
+  {
+    QFile attachmentsFile( QStringLiteral( "%1/attachments.csv" ).arg( QFieldCloudUtils::localCloudDirectory() ) );
+    QFileInfo fi( attachmentsFile );
+    if ( !fi.exists() || fi.size() == 0 )
+      return std::move( files );
+
+    attachmentsFile.open( QFile::ReadWrite | QFile::Text );
+    QTextStream attachmentsStream( &attachmentsFile );
+    while ( !attachmentsStream.atEnd() )
+    {
+      const QString line = attachmentsStream.readLine().trimmed();
+      const QStringList values = StringUtils::csvToStringList( line );
+
+      if ( values.size() >= 2 )
+      {
+        files.insert( values.at( 0 ), values.at( 1 ) );
+      }
+    }
+  }
+  return std::move( files );
+}
+
+void QFieldCloudUtils::addPendingAttachment( const QString &projectId, const QString &fileName )
+{
+  QLockFile attachmentsLock( QStringLiteral( "%1/attachments.lock" ).arg( QFieldCloudUtils::localCloudDirectory() ) );
+  if ( attachmentsLock.tryLock( 10000 ) )
+  {
+    QStringList values = QStringList() << projectId << fileName;
+
+    QFile attachmentsFile( QStringLiteral( "%1/attachments.csv" ).arg( QFieldCloudUtils::localCloudDirectory() ) );
+    attachmentsFile.open( QFile::Append | QFile::Text );
+    QTextStream attachmentsStream( &attachmentsFile );
+    attachmentsStream << Qt::endl
+                      << StringUtils::stringListToCsv( values );
+    attachmentsFile.close();
+  }
+}
+
+void QFieldCloudUtils::removePendingAttachment( const QString &projectId, const QString &fileName )
+{
+  QLockFile attachmentsLock( QStringLiteral( "%1/attachments.lock" ).arg( QFieldCloudUtils::localCloudDirectory() ) );
+  if ( attachmentsLock.tryLock( 10000 ) )
+  {
+    QStringList values = QStringList() << projectId << fileName;
+
+    QFile attachmentsFile( QStringLiteral( "%1/attachments.csv" ).arg( QFieldCloudUtils::localCloudDirectory() ) );
+    attachmentsFile.open( QFile::ReadWrite | QFile::Text );
+    QTextStream attachmentsStream( &attachmentsFile );
+    QStringList outputLines;
+    while ( !attachmentsStream.atEnd() )
+    {
+      outputLines << attachmentsStream.readLine();
+    }
+    outputLines.removeAll( StringUtils::stringListToCsv( values ) );
+    attachmentsFile.resize( 0 );
+    attachmentsStream.reset();
+    attachmentsStream << outputLines.join( '\n' );
```

Comment:
For a moment I got worried about windows when I saw `Qt::endl` above. Then I found out it is not relevant https://stackoverflow.com/a/8689547 .